kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/security/SaslAuthenticator.java

```diff
@@ -0,0 +1,265 @@
+    private AuthenticationDataSource authDataSource = null;
+    @Getter
+    private String authRole = null;
+
+    public SaslAuthenticator(PulsarService pulsarService, Set<String> allowedMechanisms) throws PulsarServerException {
+        this.authenticationService = pulsarService.getBrokerService().getAuthenticationService();
+        this.admin = pulsarService.getAdminClient();
+        this.allowedMechanisms = allowedMechanisms;
+    }
+
+    public void authenticate(RequestHeader header,
+                             AbstractRequest request,
+                             CompletableFuture<AbstractResponse> response) throws AuthenticationException {
+        switch (state) {
+            case HANDSHAKE_OR_VERSIONS_REQUEST:
+            case HANDSHAKE_REQUEST:
+                handleKafkaRequest(header, request, response);
+                break;
+            case AUTHENTICATE:
+                handleSaslToken(header, request, response);
+                break;
+            default:
+                break;
+        }
+    }
+
+    public boolean complete() {
+        return state == State.COMPLETE;
+    }
+
+    public void reset() {
+        state = State.HANDSHAKE_OR_VERSIONS_REQUEST;
+        authState = null;
+        authDataSource = null;
+        authRole = null;
+    }
+
+    private void setState(State state) {
+        this.state = state;
+        if (log.isDebugEnabled()) {
+            log.debug("Set SaslAuthenticator's state to {}", state);
+        }
+    }
+
+    private void handleKafkaRequest(RequestHeader header,
+                                    AbstractRequest request,
+                                    CompletableFuture<AbstractResponse> responseFuture) throws AuthenticationException {
+        ApiKeys apiKey = header.apiKey();
+        if (apiKey == ApiKeys.API_VERSIONS) {
+            handleApiVersionsRequest((ApiVersionsRequest) request, responseFuture);
+        } else if (apiKey == ApiKeys.SASL_HANDSHAKE) {
+            // If SaslHandshakeRequest version is v0, a series of SASL client and server tokens corresponding to the
+            // mechanism are sent as opaque packets without wrapping the messages with Kafka protocol headers.
+            // However, KoP always parses the header before a request is processed. Therefore, SaslHandshakeRequest v0
+            // shouldn't be supported, we need to send error response to client during handshake.
+            if (header.apiVersion() < 1) {
+                AuthenticationException e = new AuthenticationException("KoP doesn't support SaslHandshake v0");
+                responseFuture.complete(request.getErrorResponse(e));
+                throw e;
+            }
+            handleHandshakeRequest((SaslHandshakeRequest) request, responseFuture);
+        } else {
+            throw new AuthenticationException("Unexpected Kafka request of type " + apiKey + " during SASL handshake");
+        }
+    }
+
+    private void handleSaslToken(RequestHeader header,
```

Comment:
How about change its name to handleAuthenticate?  
token is not related with SASL.